aten/src/ATen/native/cuda/SpectralOps.cu

```diff
@@ -0,0 +1,274 @@
+
+  dst_idx_to_src_functor(const Tensor& batched_complex_signal)
+    : signal_ndim(batched_complex_signal.dim() - 1),
+      data(batched_complex_signal.data<scalar_t>()) {
+    for (int64_t i = 0; i < signal_ndim; i++) {
+      sizes[i] = batched_complex_signal.size(i);
+      strides[i] = batched_complex_signal.stride(i);
+    }
+  }
+
+  __device__ __forceinline__
+  scalar_t operator()(const int64_t& write_idx_with_imag) const
+  {
+    int64_t imag = write_idx_with_imag % 2;
+    // all but first (batch) and last (real/imag) dims need to be reflected
+    int64_t read_idx = 0;
+    int64_t remainder = write_idx_with_imag - imag;
+    int64_t dim_idx, dim_stride;
+    for (int64_t i = 0; i < signal_ndim; i++) {
+      dim_stride = strides[i];
+      dim_idx = remainder / dim_stride;
+      if (i == 0) {
+        read_idx += dim_idx * dim_stride;
+      } else if (dim_idx != 0) {
+        read_idx += (sizes[i] - dim_idx) * dim_stride;
+      }
+      remainder = remainder % dim_stride;
+    }
+    if (imag) {
+      return -data[read_idx + 1];
+    } else {
+      return data[read_idx];
+    }
+  }
+};
+
+// input should be a contiguous batched tensor of full signals (i.e. twosided)
+static inline void _fft_fill_with_hermitian_symmetry_(Tensor& input,
+                      int64_t size_last_dim, int64_t last_dim_start_slice) {
+  if (last_dim_start_slice >= size_last_dim) {
+    return;
+  }
+
+  // copy
+  int64_t n = input.numel() / size_last_dim * (size_last_dim - last_dim_start_slice);
+
+  cudaStream_t stream = globalContext().getCurrentCUDAStream();
+  auto allocator = THCThrustAllocator(globalContext().lazyInitCUDA());
+  auto policy = thrust::cuda::par(allocator).on(stream);
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(input.type(), "_fft_fill_with_hermitian_symmetry_", [&] {
+    using cuda_scalar_t = cuda::type<scalar_t>;
+    typedef thrust::device_ptr<cuda_scalar_t> device_ptr;
+    typedef thrust::counting_iterator<int64_t> counter;
+    typedef thrust::transform_iterator<cnt_to_dst_idx_functor, counter> dst_idx_iterator;
+    typedef thrust::permutation_iterator<device_ptr, dst_idx_iterator> dst_iterator;
+    typedef thrust::transform_iterator<dst_idx_to_src_functor<cuda_scalar_t>, dst_idx_iterator> src_iterator;
+
+    dst_idx_iterator dst_idxs(counter(0), cnt_to_dst_idx_functor(size_last_dim, last_dim_start_slice));
+
+    auto data = device_ptr(input.data<cuda_scalar_t>());
+    dst_iterator dsts(data, dst_idxs);
+    src_iterator srcs(dst_idxs, dst_idx_to_src_functor<cuda_scalar_t>(input));
+    thrust::copy_n(policy, srcs, n, dsts);
+  });
+}
+
+// cufft
+Tensor _fft_cufft(const Tensor& self, int64_t signal_ndim,
+                  bool complex_input, bool complex_output,
+                  bool inverse, IntList checked_signal_sizes,
+                  bool normalized, bool onesided,
+                  IntList output_sizes) {
+  Tensor input = self;
+
+  bool is_half = input.type().scalarType() == ScalarType::Half;
```

Comment:
I am reading the half restrictions and they state "More than one GPU is not supported". Do you know what this means?

Reply:
Also, it states minimum GPU architecture is `SM_53`; I don't see a test for this anywhere.

Reply:
Re "More than one GPU is not supported", this appears to apply to uses of `cufftXtSetGPUs`, which is not used in this PR. It's worth noting somewhere that our cufft implementation doesn't make use of multiple gpus (so it could be faster.)

Reply:
(Also, it's more annoying to implement, because you need to do `cufftXtMemcpy` to set this up.)

Reply:
Added a test for this.